critiquebrainz/db/review.py

```diff
@@ -0,0 +1,577 @@
+            raise db_exceptions.BadDataException("Changing license of a published review is not allowed.")
+        updates.append("license_id = :license_id")
+        updated_info["license_id"] = license_id
+
+    if language is not None:
+        updates.append("language = :language")
+        updated_info["language"] = language
+
+    if is_draft is not None:
+        if not drafted and is_draft: # If trying to convert published review into draft
+            raise db_exceptions.BadDataException("Converting published reviews back to drafts is not allowed.")
+        updates.append("is_draft = :is_draft")
+        updated_info["is_draft"] = is_draft
+
+    setstr = ", ".join(updates)
+    query = sqlalchemy.text("""
+        UPDATE review
+           SET {setstr}
+         WHERE id = :review_id
+    """.format(setstr=setstr))
+
+    if setstr:
+        updated_info["review_id"] = review_id
+        with db.engine.connect() as connection:
+            connection.execute(query, updated_info)
+    # Create new revision
+    new_revision = db_revision.create(review_id, text)
+    cache.invalidate_namespace(REVIEW_CACHE_NAMESPACE)
+
+
+def create(*, entity_id, entity_type, user_id, is_draft, text,
+           language=DEFAULT_LANG, license_id=DEFAULT_LICENSE_ID,
+           source=None, source_url=None):
+    """Create a new review on an entity.
+    Args:
+        entity_id (uuid): ID of the entity to be reviewed.
+        entity_type (str): Type of the entity reviewed.
+        user_id (uuid): ID of the reviewer.
+        is_draft (bool): Whether the review is drafted.
+        text (str): Content of the review.
+        language (str): Language of the review written. (default: 'en')
+        license_id (str): ID of the license.
+        source (str): Source of the review.
+        source_url (str): Url of the source of the review.
+
+    Returns:
+        Dictionary with the following structure
+        {
+            "id": uuid,
+            "entity_id": uuid,
+            "entity_type": str("release group", "event", "place"),
+            "user_id": uuid,
+            "user": dict,
+            "edits": int,
+            "is_draft": bool,
+            "is_hidden": bool,
+            "language": str,
+            "license_id": str,
+            "source": str,
+            "source_url": str,
+            "last_revision: dict,
+            "votes": dict,
+            "rating": int,
+            "text": str,
+            "created": datetime,
+            "license": dict,
+        }
+    """
+    if language not in supported_languages:
+        raise ValueError("Language: {} is not supported".format(language))
+
+    with db.engine.connect() as connection:
+        result = connection.execute(sqlalchemy.text("""
+            INSERT INTO review
+            VALUES (:id, :entity_id, :entity_type, :user_id, :edits, :is_draft,
+            :is_hidden, :license_id, :language, :source, :source_url)
```

Comment:
This way of insertion is unsafe and caused CB-258. Order of columns needs to be explicitly specified. Here you are making an assumption that `entity_type` is third, which it might be when a new instance of CB database is created, but it isn't in the production database.